components/ai_chat/core/browser/tab_tracker_service.cc

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ai_chat/core/browser/tab_tracker_service.h"
+
+#include <algorithm>
+#include <iterator>
+#include <utility>
+
+#include "brave/components/ai_chat/core/common/constants.h"
+#include "brave/components/ai_chat/core/common/mojom/tab_tracker.mojom.h"
+
+namespace ai_chat {
+
+TabTrackerService::TabTrackerService() = default;
+TabTrackerService::~TabTrackerService() = default;
+
+void TabTrackerService::Bind(
+    mojo::PendingReceiver<mojom::TabTrackerService> receiver) {
+  receivers_.Add(this, std::move(receiver));
+}
+
+void TabTrackerService::UpdateTab(int32_t tab_id, mojom::TabDataPtr tab) {
+  auto allowed_scheme =
+      tab && base::Contains(kAllowedContentSchemes, tab->url.scheme());
```

Comment:
there are several checks for `tab`, would it ever be null and if it is, wouldn't we want to just return early at the start?

Reply:
No, because setting a `tab_id` to `nullptr` means the tab has been deleted (so we should remove it from out list of tracked tabs).